### PR TITLE
feat(treesitter): highlight help files by default

### PR DESCRIPTION
**Pull request by clason**
I think it's safe to enable this by default for 0.10 so everyone can appreciate our shiny syntax-highlighted code blocks.

**Question:** Do we want to provide an opt-out?

### CODE REVIEW

**Review by lewis6991 — APPROVED** [review]

**Comment by seandewar** [comment]
May need to add the `build/lib/nvim` dir to the &rtp in runtest.vim, as that's where the parsers are built to.

As for why it works without `:filetype plugin on`, it's because the runner uses `-u NONE`, so all the filetype detection stuff is usually off.

**Comment by wookayin** [comment]
Is there any way to disable it or opt-out by any chance? I'm not quite sure but I was wondering about the case where `vim.treesitter.start()` throws an error due to invalid/outdated parser and incompatible query files (which can conflict with the parsers compiled and managed by nvim-treesitter).

**Comment by justinmk** [comment]
> Do we want to provide an opt-out?

Isn't `vim.treesitter.stop()` the opt-out?

**Comment by clason** [comment]
> Isn't `vim.treesitter.stop()` the opt-out?

Not quite; enabling and then disabling is not exactly the same as skipping enabling. (For example, you'd still get errors if `vim.treesitter.start()` doesn't work for some reason.) Not saying we _need_ to provide an opt-out, mind you.

**Comment by justinmk** [comment]
> not exactly the same as skipping enabling.

How is that normally done for an ftplugin or syntax plugin?

**Comment by clason** [comment]
It is not, really.

**Comment by Shougo** [comment]
```lua
vim.treesitter.start = (function(wrapped)
  return function(bufnr, lang)
    if lang == nil then
      return
    end
    wrapped(bufnr, lang)
  end
end)(vim.treesitter.start)
```

It works for me.

File: runtime/ftplugin/help.lua
Comment (Shougo) [inline runtime/ftplugin/help.lua:2]: `lang = nil` is passed.
I think lang should be `vimdoc`.

Comment (Shougo) [reply runtime/ftplugin/help.lua:2]: https://github.com/neovim/neovim/pull/26347#issuecomment-1837508178

It does not work if the language is auto detected.

**Comment by justinmk** [comment]
> monkey patch is bad idea because sometime happens conflicts with other changes.

What other changes? 

> `lang` is nil when help file. Oh no.

I updated the code in https://github.com/neovim/neovim/pull/26347#issuecomment-1837508178 to also check 'filetype'.

**Comment by kuuote** [comment]
I apologize for providing my opinion without reading the documents.
If `vim.treesitter.start()` is dedicated for highlighting purposes, then indeed, there is no need to worry about conflicts.

**Comment by Shougo** [comment]
> I updated the code in https://github.com/neovim/neovim/pull/26347#issuecomment-1837508178 to also check 'filetype'.

Thank you.  It works.

**Comment by kyoh86** [comment]
@justinmk @clason 
Is the conclusion of this discussion that the monkey patch is the "only" solution going forward?
I hope that an option (not a monkey patch) will be provided in the future to suppress treysitter starts.

**Comment by clason** [comment]
No, that is not planned. The goal is to make treesitter a safe and, all things considered, better choice.

**Comment by kyoh86** [comment]
@clason 
I see.
I understand your premise about the incompleteness of treesitter, which can be improved in the future.
However, the cost of treesitter will never be zero, just as the old syntax is.
I would appreciate it if you could provide users with a straightforward way to eliminate this, if possible.

**Comment by clason** [comment]
I hear you, but the answer is still no. There should be no (reasonable) reason not to have this enabled, so maintaining such an option is wasted effort that is better spent elsewhere.

**Comment by kyoh86** [comment]
@clason 
Isn't it REASONABLE that users who don't use treesitter don't want to pay that cost?

**Comment by clason** [comment]
No, because "that cost" is not well defined. You're not complaining about the legacy regex syntax engine being enabled by default, and that one is not "free" either. So why should treesitter be different?

**Comment by kyoh86** [comment]
@clason 
Sure, I could agree that "that cost" is not well defined.
If I could define well it, I may discuss about it again (in another issue).

However, we can disable the legacy regex syntax engine by straight way (`:syntax off`).
I don't think it is such an outlandish request to provide a method that is as simple as it is.

**Comment by clason** [comment]
But `vim.treesitter.stop()` already exists. That's the exact analogue of `:syntax off`.

**Comment by kyoh86** [comment]
I see.
I agree that "last resort" it would be an easy solution for the maintainer, but whether it is a "straight" way or not is subjective, so I will defer to your wishes for once.

Thanks for answering so many of my questions.

**Pull request merged**
